### ISSUE TITLE
Colab can't import trim_batch, anything changed in transformers libraries?

### ISSUE THREAD

**Issue by yxu1168** [enzoampil/t5-intro#1]
from transformers.tokenization_utils import trim_batch

ImportError: cannot import name 'trim_batch'

Any solutions? Thanks.

**Comment by WadoodAbdul** [comment]
You can remove that line and add this function

`def trim_batch(
    input_ids, pad_token_id, attention_mask=None,
):
    """Remove columns that are populated exclusively by pad_token_id"""
    keep_column_mask = input_ids.ne(pad_token_id).any(dim=0)
    if attention_mask is None:
        return input_ids[:, keep_column_mask]
    else:
        return (input_ids[:, keep_column_mask], attention_mask[:, `keep_column_mask])`

**Comment by WadoodAbdul** [comment]
You can remove that line and add this function

`def trim_batch(input_ids, pad_token_id, attention_mask=None,):
     keep_column_mask = input_ids.ne(pad_token_id).any(dim=0) 
     if attention_mask is None: return input_ids[:, keep_column_mask] 
     else: return (input_ids[:, keep_column_mask], attention_mask[:, keep_column_mask])`

[https://github.com/huggingface/transformers/blob/1e51bb717c04ca4b01a05a7a548e6b550be38628/src/transformers/tokenization_utils.py](url)

**Comment by yxu1168** [comment]
Thanks a lot.

**Comment by WadoodAbdul** [comment]
If you are trying to run the colab notebook, you'll run across more errors after fixing this.
If yes, hit me up, I have solved a few, we can fix the rest together.

**Comment by enzoampil** [comment]
Thanks for resolving this @WadoodAbdul :smile:

Apologies saw this one late @yxu1168

**Comment by yxu1168** [comment]
Thanks @WadoodAbdul.
After fix the trim_batch issue, I was able to run the code on Colab.
In the finetune part, I didn't change any hyperparameters, just run the original code, the acc: jaccard_score = 0.503
is very low. Each epoch took more than half hours. For the test Q/A part, the answer wasn't very good.
The previous thought to run this notebook was to see if I can implement T5 to the COVID-19 QA. Any suggestions?

________________________________

**Comment by WadoodAbdul** [comment]
@enzoampil No problem. Thanks for your amazing talk on Abhishek Thakur's channel!

@yxu1168 Did you use GPU for training?
I had to add some additional arguments in the optimizer_step function for the code to work. The training part was quick but there was no improvement in the results across the epochs.
Regarding the COVID-19 QA dataset, I think squad performs well on question-answering tasks too. To implement T5 for QA you'd have to process the inputs and targets according to the format metioned in section D.15 in the [paper](https://arxiv.org/pdf/1910.10683.pdf).

**Comment by yxu1168** [comment]
@WadoodAbdul<mailto:notifications@github.com>
Thanks for the quick response and the link of the paper, I'll study it.
It would be nice to run fast then we can increase epochs to see what happen. Have you updated your notebook in your Github repo? If so, can I search your Github account by WadoodAbdul<mailto:notifications@github.com>?
Yes I used Colab GPU.
By the way, we use below cell to fine tune lr, batch size, epochs right? I am not familiar to linux.
ARGS_STR =
--data_dir=/content/ \
--model_name_or_path=t5-base \
--learning_rate=3e-5 \
--train_batch_size=32 \
--output_dir=/content/output/ \
--do_train \
--n_gpu=1 \
--num_train_epochs 5 \
--max_source_length 80 \

Compare with BERT, GPT-2/3,  do you think T5 combines the advantages from both?
Basically, I'd like to understand T5 and this notebook well, and how good it works to solve NLP problems by transfer learning, etc.
Regards,


________________________________

**Comment by WadoodAbdul** [comment]
@yxu1168 

(i) Since we are using pytorch-lightning you could use TPUs for faster computation.

(ii) I haven't updated the notebook in a github repo, I'll try fixing the error I'm facing by today.

(iii) Yes,this is the cell where we pass the fine tuning arguments.

(iv) T5 combines the architecture from both BERT and GPT2/3, I'm not sure about the advantages though :).
It does show a lot of promise(inferred from its results). I too am in the process of understanding this stuff, I'll share my notebook once I get a good understanding of the structure.

**Comment by yxu1168** [comment]
@WadoodAbdul
Owesome, thank you very much, looking forward to hearing from you for the progress soon.
Regards,


________________________________
